### ISSUE TITLE
amplify push fails on "The specified key does not exist ... Error Code: NoSuchKey;"

### ISSUE THREAD

**Issue by parvusville** [aws-amplify/amplify-cli#8236]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### How did you install the Amplify CLI?

npm

### Amplify CLI Version

6.0.1

### What operating system are you using?

Pop Os

### Amplify Categories

auth, storage, function, api

### Amplify Commands

push

### Describe the bug

Amplify push fails on `CREATE_FAILED XXXResolver AWS::AppSync::Resolver Tue Sep 21 2021 21:53:45 GMT+0300 (Eastern European Summer Time) The specified key does not exist. (Service: Amazon S3; Status Code: 404; Error Code: NoSuchKey; Request ID: XXX; S3 Extended Request ID: XXX=; Proxy: null)`

### Expected behavior

Push should succeed

### Reproduction steps

I have been merging changes from dev to staging, and accidentally did amplify push to dev environment while I had staging codebase open. I also updated to Amplify CLI 6.0.1 earlier today, but I'm pretty sure I did successful pushes after that. 

### GraphQL schema(s)

<details>

```graphql
# Put schemas below this line


```

</details>

### Log output

<details>

```
# Put your logs below this line


```

</details>

### Additional information

I have verified that the deployment bucket for dev environment exists in S3. Pushed work fine in staging environment.

**Comment by attilah** [comment]
@parvusville based on the description I am sure that you run into something that is not even supposed to mean, but to confirm that could you provide exact repro steps and more details about what resource failed only 1 resolver? I'd like to close out that we are having a valid issue.

Is that a resolver that is for a type which is different between `staging` and `dev`?

**Comment by parvusville** [comment]
@attilah Ah. In dev I have this type where this XXX field was the only one missing @auth directive. In staging as a part of getting push through I had to remove all of those @auths and then added them back in the next push (including this XXX field). I  added the @auth for the field in dev too and now pushing works again :) I realized this when I was already done writing the steps I took,  so leaving them here anyways in case you can use this for something..:

I am not 100% sure, but this is how **I think**  it happened. 
1: Before these problems I merged changed from dev to staging and pushed. Including changes to GraphQL API, new REST API and multiple new functions
2: CICD fails because of this https://github.com/aws-amplify/amplify-cli/issues/6231
3: Fixing previous by removing @auth rules from non-model types and pushing, CICD now fails because of some one the new functions had environment and secret variables set up.
4: Now I did amplify env checkout staging and amplify push. It asks me to set up the variables and goes on.
5: Push fails on https://github.com/aws-amplify/amplify-cli/issues/4094 . 
6: I removed the index from dynamodb console
7: amplify push works now
8: amplify env checkout dev
I notice that in staging some graphql requests aren't going through because of removing @auth rules at step 3.
9: I add @rules back in staging git branch. I'm not 100% sure if I used git and CI/CD or amplify push. Probably I used CICD and everything still worked fine
Now I believe im in state where amplify env dev is active, and staging git branch is active. 
10:  amplify function update, give function Resource access permission for my REST API. (I wanted the API_MYAPI_APIID environment variable to see if I can build REST url dynamically for each enivornment)
`UPDATE_COMPLETE    GraphQLAPIKey                                   AWS::AppSync::ApiKey       Tue Sep 21 2021 21:02:24 GMT+0300 (Eastern European Summer Time)                                                                                                                         
CREATE_FAILED      XXXResolver AWS::AppSync::Resolver     Tue Sep 21 2021 21:02:24 GMT+0300 (Eastern European Summer Time) The specified key does not exist. (Service: Amazon S3; Status Code: 404; Error Code: NoSuchKey; Request ID: XXX; S3 Extended Request ID: XXX; Proxy: null)
⠹ Updating resources in the cloud. This may take a few minutes...

UPDATE_ROLLBACK_IN_PROGRESS XXX-123-myApi-XXX AWS::CloudFormation::Stack Tue Sep 21 2021 21:02:26 GMT+0300 (Eastern European Summer Time) The following resource(s) failed to create: [XXXResolver].
⠦ Updating resources in the cloud. This may take a few minutes...

UPDATE_FAILED myApi AWS::CloudFormation::Stack Tue Sep 21 2021 21:02:35 GMT+0300 (Eastern European Summer Time) Embedded stack arn:aws:cloudformation:eu-west-1:1234:stack/amplify-myApp-dev-115515-myApp-123/1234-5588-11eb-9903-021248ca0799 was not successfully updated. Currently in UPDATE_ROLLBACK_IN_PROGRESS with reason: The following resource(s) failed to create: [UpsertReservationResultdestinationEmailResolver].
⠋ Updating resources in the cloud. This may take a few minutes...

UPDATE_ROLLBACK_IN_PROGRESS amplify-urentdealer-dev-115515 AWS::CloudFormation::Stack Tue Sep 21 2021 21:02:36 GMT+0300 (Eastern European Summer Time) The following resource(s) failed to update: [apiurentdealer].
⠸ Updating resources in the cloud. This may take a few minutes...`

**Comment by attilah** [comment]
I'm glad that you partially unblocked yourself as I see a lot is going on.

This error: `The specified key does not exist` most probably coming from the fact that your deployment bucket does not have the files for the resolver: `UpsertReservationResultdestinationEmailResolver` could you verify if that is the case? also check if you see it in the `build` folder for your API resource and confirm that this resolver is something that you'd expect to be present based on your schema.

**Comment by parvusville** [comment]
As far as I know everything works now after I added the @auth to the destinationEmail field in dev and pushed it. In the build folder I now have UpsertReservationResult.destinationEmail.req/res.vtl, and also in the #current-cloud-backend.zip in the deployment bucket. I downloaded that folder also when I had problem with the push, and indeed in that version I was missing that resolver.
